internal/controller/object/informers.go

```diff
@@ -0,0 +1,273 @@
+		if found {
+			continue
+		}
+
+		log := i.log.WithValues("providerConfig", providerConfig, "gvk", gvk.String())
+
+		ca, err := cache.New(rc, cache.Options{})
+		if err != nil {
+			log.Debug("failed creating a cache", "error", err)
+			continue
+		}
+
+		// don't forget to call cancelFn in error cases to avoid leaks. In the
+		// happy case it's called from the go routine starting the cache below.
+		ctx, cancelFn := context.WithCancel(context.Background())
+
+		u := kunstructured.Unstructured{}
+		u.SetGroupVersionKind(gvk)
+		inf, err := ca.GetInformer(ctx, &u, cache.BlockUntilSynced(false)) // don't block. We wait in the go routine below.
+		if err != nil {
+			cancelFn()
+			log.Debug("failed getting informer", "error", err)
+			continue
+		}
+
+		if _, err := inf.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				ev := runtimeevent.GenericEvent{
+					Object: obj.(client.Object),
+				}
+
+				i.sink(providerConfig, ev)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				old := oldObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				obj := newObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				if old.GetResourceVersion() == obj.GetResourceVersion() {
+					return
+				}
+
+				ev := runtimeevent.GenericEvent{
+					Object: obj,
+				}
+
+				i.sink(providerConfig, ev)
+			},
+			DeleteFunc: func(obj interface{}) {
+				ev := runtimeevent.GenericEvent{
+					Object: obj.(client.Object),
+				}
+
+				i.sink(providerConfig, ev)
+			},
+		}); err != nil {
+			cancelFn()
+			log.Debug("failed adding event handler", "error", err)
+			continue
+		}
+
+		go func() {
+			defer cancelFn()
+
+			log.Info("Starting resource watch")
+			_ = ca.Start(ctx)
+		}()
+
+		i.lock.Lock()
+		i.resourceCaches[gvkWithConfig{providerConfig: providerConfig, gvk: gvk}] = resourceCache{
+			cache:    ca,
+			cancelFn: cancelFn,
+		}
+		i.lock.Unlock()
+
+		// wait for in the background, and only when synced add to the routed cache
+		go func() {
+			if synced := ca.WaitForCacheSync(ctx); synced {
+				log.Debug("Resource cache synced")
+			}
+		}()
+	}
+}
+
+func (i *resourceInformers) StopWatchingResources(ctx context.Context, providerConfig string, gvks ...schema.GroupVersionKind) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	for _, gvk := range gvks {
+		ca, found := i.resourceCaches[gvkWithConfig{providerConfig: providerConfig, gvk: gvk}]
+		if !found {
+			continue
+		}
+		// Check if there are any other objects referencing this resource GVK.
+		list := v1alpha2.ObjectList{}
+		if err := i.objectsCache.List(ctx, &list, client.MatchingFields{resourceRefGVKsIndex: refKeyProviderGVK(providerConfig, gvk.Kind, gvk.Group, gvk.Version)}); err != nil {
+			i.log.Debug("cannot list objects referencing a certain resource GVK", "error", err, "fieldSelector", resourceRefGVKsIndex+"="+refKeyProviderGVK(providerConfig, gvk.Kind, gvk.Group, gvk.Version))
```

Comment:
return

Reply:
continue indeed since we're looping input gvks